// arch-typescript-sdk/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Status = exports.Pubkey = void 0;
const buffer_1 = require("buffer");
class Pubkey {
    constructor(bytes) {
        this.bytes = bytes;
        if (bytes.length !== 32) {
            throw new Error('Pubkey must be 32 bytes');
        }
    }
    static fromString(s) {
        const bytes = buffer_1.Buffer.from(s, 'hex');
        if (bytes.length !== 32) {
            throw new Error('Invalid pubkey string');
        }
        return new Pubkey(bytes);
    }
    toString() {
        return buffer_1.Buffer.from(this.bytes).toString('hex');
    }
    serialize() {
        return Array.from(this.bytes);
    }
    static systemProgram() {
        const bytes = new Uint8Array(32);
        bytes[31] = 1;
        return new Pubkey(bytes);
    }
}
exports.Pubkey = Pubkey;
var Status;
(function (Status) {
    Status[Status["Processing"] = 0] = "Processing";
    Status[Status["Processed"] = 1] = "Processed";
})(Status || (exports.Status = Status = {}));
